staging/src/k8s.io/apiserver/pkg/apis/apiserver/validation/validation.go

```diff
@@ -334,24 +339,83 @@ func ValidateWebhookConfiguration(fldPath *field.Path, c *api.WebhookConfigurati
 		allErrs = append(allErrs, field.NotSupported(fldPath.Child("connectionInfo", "type"), c.ConnectionInfo, []string{api.AuthorizationWebhookConnectionInfoTypeInCluster, api.AuthorizationWebhookConnectionInfoTypeKubeConfigFile}))
 	}
 
-	// TODO: Remove this check and ensure that correct validations below for MatchConditions are added
-	// for i, condition := range c.MatchConditions {
-	//	 fldPath := fldPath.Child("matchConditions").Index(i).Child("expression")
-	//	 if len(strings.TrimSpace(condition.Expression)) == 0 {
-	//	     allErrs = append(allErrs, field.Required(fldPath, ""))
-	//	 } else {
-	//		 allErrs = append(allErrs, ValidateWebhookMatchCondition(fldPath, sampleSAR, condition.Expression)...)
-	//	 }
-	// }
-	if len(c.MatchConditions) != 0 {
-		allErrs = append(allErrs, field.NotSupported(fldPath.Child("matchConditions"), c.MatchConditions, []string{}))
-	}
+	_, errs := validateMatchConditions(c.MatchConditions, fldPath, utilfeature.DefaultFeatureGate.Enabled(features.StructuredAuthorizationConfiguration))
+	allErrs = append(allErrs, errs...)
 
 	return allErrs
 }
 
-func ValidateWebhookMatchCondition(fldPath *field.Path, sampleSAR runtime.Object, expression string) field.ErrorList {
-	allErrs := field.ErrorList{}
-	// TODO: typecheck CEL expression
-	return allErrs
+// ValidateMatchConditions validates a given webhook's matchConditions.
+// This is exported for use in authz package.
+func ValidateMatchConditions(matchConditions []api.WebhookMatchCondition) (*authorizationcel.CELMapper, field.ErrorList) {
+	return validateMatchConditions(matchConditions, nil, utilfeature.DefaultFeatureGate.Enabled(features.StructuredAuthorizationConfiguration))
+}
+
+func validateMatchConditions(matchConditions []api.WebhookMatchCondition, fldPath *field.Path, structuredAuthzFeatureEnabled bool) (*authorizationcel.CELMapper, field.ErrorList) {
```

Comment:
Why do we need to keep passing the feature gate?

Reply:
Each webhook in the config needs to validate its matchConditions against the feature gate. The feature only needs to be enabled if more than one matchCondition is defined. Otherwise, the feature gate is not needed. Did you have other suggestions?

Reply:
checking the feature flag in validateMatchConditions ensures we have the same validation logic and the same error returned. Doing it before https://github.com/kubernetes/kubernetes/pull/121223/files#diff-89d23554d408bdc1a1f23594bcc817a77f1532cf8390a4ba179b33a17e5ed845R342 and https://github.com/kubernetes/kubernetes/pull/121223/files#diff-8b41b5e2c7d3bb27fa452e58ab0a4120b02d3a7a16d8961954f5c3c270ab8935R115, would require us to ensure both places are updated exactly the same way to ensure consistency.